### PR TITLE
Docs Versioning++

### PR DESCRIPTION
**Pull request by JKarlavige**
## Description & motivation
This introduces versioning to the docs site. This does not use the built-in versioning Docusaurus includes as we do not want multiple duplicates created for different versions of files. This instead uses a global context to set the version and a component to determine if the item within a docs page should show.

## Pre-release docs
Is this change related to an unreleased version of dbt?
- [ ] Yes: please
    - update the base branch to `next`
    - add Changelog components: `<Changelog>[New/Changed] in v0.x.0</Changelog>`
    - add links to the "New and changed documentation" section of the latest [Migration Guide](../website/docs/docs/guides/migration-guide)
- [x] No: please ensure the base branch is `current`
- [ ] Unsure: we'll let you know!

## Preview
https://deploy-preview-1052--docs-getdbt-com.netlify.app/reference/node-selection/defer

### CODE REVIEW

File: website/docs/reference/node-selection/defer.md
Comment (joellabes) [inline website/docs/reference/node-selection/defer.md]: does this imply that firstVersion is a `>=` and lastVersion is a `<=`? There could be behavioural changes in patches as well, are we choosing to not support that? If so then I'd advocate for the headers being 0.21.x, 1.0.x, etc

**Comment by joellabes** [comment]
Will we store a user's selected version as a cookie?

**Comment by JKarlavige** [comment]
> Will we store a user's selected version as a cookie?

That's a good question @joellabes, I think we should. At the moment the version is lost once a user refreshes the page or leaves the site. I added it to the list in Asana to get this done.

File: website/docs/reference/node-selection/defer.md
Comment (JKarlavige) [reply website/docs/reference/node-selection/defer.md]: Yes that's correct, although this is an early prototype so far to test how it will work. The `lastVersion` prop will be optional, so that value should only be set if a feature no longer exists but we still want to show it for an older version in the docs. This will prevent us from having to update all lastVersion props throughout the docs each time we add a new version to the site.

You make a good point regarding the patches. If a feature was released on version 0.20.3 for example, and we only have v0.20.0 and v1.0.2 in the versions dropdown, the user would have to select v1.0.2 to view that section. I think we should discuss this with @runleonarun to start narrowing down specifics on what versions we will have on the site, whether we should support patches, etc.

Comment (runleonarun) [reply website/docs/reference/node-selection/defer.md]: I believe when @jtcohen6 and I talked about this, we decided to document major (1.x) versions only. The idea being that minor versions would be bug fixes only that should not normally affect docs. I think 1.x is the furthest we'd need to slice the docs, right Jeremy?

Comment (jtcohen6) [reply website/docs/reference/node-selection/defer.md]: @runleonarun Right on, with just the minor clarification that `1.x` = "minor" and `1.x.y` = "patch".

To that end:

> There could be behavioural changes in patches as well, are we choosing to not support that? If so then I'd advocate for the headers being 0.21.x, 1.0.x, etc

IMO any behavioral changes in patches would restore/represent the "correct" behavior for that entire minor version family. If we change something in v1.0.2, it's to _correct_ the behavior that everyone using v1.0 was meant to experience. As an operating principle, I think we should assume that everyone using v1.0 is / ought to be using the latest patch for v1.0.

> If a feature was released on version 0.20.3 for example

- We will not be releasing any new features in patch versions (such as v0.20.3), only minor versions (v0.21.0)
- We will not be releasing any new versions pre-v1.0

Comment (JKarlavige) [reply website/docs/reference/node-selection/defer.md]: Thanks for the details @jtcohen6, helpful to know what is expected from minor and patch releases. From both and editor and development standpoint, excluding patches from the versioning on the site will reduce the complexities of having to manage the versions overall.

It would be easier to maintain and add new blocks of versioned content, while also simplifying the logic quite a bit to check if the feature is available for the current selected version.

File: website/src/theme/DocPage/index.js
Comment (jtcohen6) [inline website/src/theme/DocPage/index.js]: In addition to warning people that the page is unavailable for their current version, could we also tell them the first version in which the feature does become available? Then we get to frame this more positively:
```
🎉 New feature!
Unfortunately, it's not available in ${dbtVersion}
Upgrade to ${firstVersion} or later to use it
```

File: website/src/theme/DocPage/styles.module.css
Comment (jtcohen6) [inline website/src/theme/DocPage/styles.module.css]: With the needed caveat that I am very much not a designer: I was expecting something in the yellow/orange/red color family, to contrast more with the standard menu color

File: website/sidebars.js
Comment (jtcohen6) [inline website/sidebars.js:634]: Whoa, versioned sidebar?? Fancy!

File: website/docs/reference/node-selection/defer.md
Comment (jtcohen6) [inline website/docs/reference/node-selection/defer.md]: Nice!! This works intuitively and is super easy to use.

The only thing that doesn't seem to work is specifying a `lastVersion` only. I don't know if/when we'd ever actually want to do that. Perhaps limitations, caveats, or known bugs that can be removed when we've resolved them, and no longer bear mentioning? Even so, it's easy enough to just write as `<VersionBlock firstVersion="0.0" lastVersion="...">`

File: website/dbt-versions.js
Comment (jtcohen6) [inline website/dbt-versions.js]: In case others are confused at first, as I was:
- These are for pages documenting functionality are _entirely_ unsupported in older (or newer) versions of dbt
- It's not necessary to add every page to this list, every time there's a small difference in functionality across versions. That's what the `<VersionBlock>` is for

File: website/dbt-versions.js
Comment (jtcohen6) [inline website/dbt-versions.js]: Copying from Slack:

> I'm not strictly opposed to including v0.20 + v0.21, especially since we already have those modern/legacy tabs for the `--models`/`--select` rename. I wonder if we can do that, while still making clear that only v1.0 and newer are "officially supported" these days? Perhaps by showing a banner when a user selects v0.20 or v0.21? "You're viewing docs for an outdated version of dbt"?

Context: Starting with v1.0, a minor version of dbt will be "officially supported" for 12 months from its initial release. All pre-v1.0 versions are past their official support period. The new Core Versions page (#1061) will also help with clarifying + setting these expectations.

I could see us adding "end of life" dates to these version constants, and then warning people when looking at docs for a version that:
- has an EOL date fast approaching (say, within the next 3 months)
- is already past its official support period

@runleonarun @JKarlavige What do you think of showing those types of banners for _all_ pages? I think it could go a long way toward teaching users about our version policies. I'm happy to kick that further down the road, and say it's out of scope for this PR. Such a banner also ought to link people to the "Understanding dbt Core versions" page, which is soon to exist :)

**Review by jtcohen6 — COMMENTED** [review]
This is fantastic work! So excited about what's already here, and about the next steps. I left a handful of comments, none of which is blocking

File: website/dbt-versions.js
Comment (jtcohen6) [reply website/dbt-versions.js]: How [Jinja](https://jinja.palletsprojects.com/en/2.11.x/) does this, for instance:

Screenshot 2022-02-21 at 13 15 46

File: website/src/theme/DocPage/index.js
Comment (JKarlavige) [reply website/src/theme/DocPage/index.js]: I think this is a great way to frame it! I will update the banner with this content.

File: website/src/theme/DocPage/styles.module.css
Comment (JKarlavige) [reply website/src/theme/DocPage/styles.module.css]: @runleonarun and I looked at some color options based on our brand colors. I agree we should adjust this to find a way to make it more of an alert, and separate from a standard note.

File: website/docs/reference/node-selection/defer.md
Comment (JKarlavige) [reply website/docs/reference/node-selection/defer.md]: At the moment, the `firstVersion` is required and the `lastVersion` is optional. Rather than having to set `firstVersion="0.0"`, it would be a better experience to be able to exclude the firstVersion prop altogether. That should definitely be possible, so I will revisit this to adjust

File: website/dbt-versions.js
Comment (JKarlavige) [reply website/dbt-versions.js]: That is spot on, thanks for noting that here for others looking over it. I have an outline of documentation for the versioning [in progress here](https://www.notion.so/dbtlabs/Versioning-on-Docusaurus-c6a4a41a66cd4ea2970854cc42cb5b70), and these notes will definitely be great to have in there as well.

File: website/dbt-versions.js
Comment (JKarlavige) [reply website/dbt-versions.js]: This would be a great feature to have, and i'm open to getting it added into this PR. Docusaurus has a banner for this which works with their built-in versioning. [Example here](https://61c1eade2ffed20008aaac30--docusaurus-2.netlify.app/docs/2.0.0-beta.10). 

Since we are already showing a banner for the versioned pages if they are not available for that version, it should be straight forward enough to adjust the versions const to add an "end of life" property for each version. If the end of life date has passed, we can show the alert.

In the meantime, it would be great to sort out the language we would like to use for this unsupported version banner.

Comment (runleonarun) [reply website/dbt-versions.js]: > In the meantime, it would be great to sort out the language we would like to use for this unsupported version banner.

I can work on this!

File: website/dbt-global-variables.js
Comment (joellabes) [inline website/dbt-global-variables.js]: @runleonarun what sorts of things do we think we'll be defining in here? At EP we had a Brand variable for our automated emails and it wound up being more of a pain than a benefit, because we did a bad job of guessing what components were going to be renamed over time.

Made up example - if we had `dbtCloud` as a catchall variable for the SaaS product, and then we come up with a fun name for the IDE like `The Power Console` (🤢 ) then we still have to go and split out the new component into its smaller chunks... 

Slightly concerned at the thought that we might have recency bias around "what if the company name changes again" which will hopefully not happen again any time soon! 

100% open to being talked out of it if you've found them useful in a past life though

**Review by joellabes — COMMENTED** [review]
I spent some more time messing around today. Some thoughts: 
- it's weird to me that the [metrics page disappears from the sidebar](https://deploy-preview-1052--docs-getdbt-com.netlify.app/docs/building-a-dbt-project/metrics) when I'm using an old version - it feels like my breadcrumbs are broken. 
- The version data isn't in the URL anywhere, which makes it impossible to share version-specific docs over Slack/support tickets 
- Design-wise, it's weird to have the version picker on the right hand sidebar, which is the jumping off point to _leave_ the site (dev blog, dbt learn, community links...)

**Comment by JKarlavige** [comment]
> I spent some more time messing around today. Some thoughts:
> 
> * it's weird to me that the [metrics page disappears from the sidebar](https://deploy-preview-1052--docs-getdbt-com.netlify.app/docs/building-a-dbt-project/metrics) when I'm using an old version - it feels like my breadcrumbs are broken.
> * The version data isn't in the URL anywhere, which makes it impossible to share version-specific docs over Slack/support tickets
> * Design-wise, it's weird to have the version picker on the right hand sidebar, which is the jumping off point to _leave_ the site (dev blog, dbt learn, community links...)

Thanks for the feedback Joel, i'll add more context for each here:

- The sidebar items hide entirely if they're not available for the current version. However, I think it's still very much an open discussion on whether we want to hide them entirely, or if the banner which shows above the page would be enough
- In order to share a version-specific url, this can be done by adding a `version` param to the url, for example: https://deploy-preview-1052--docs-getdbt-com.netlify.app/?version=0.21
- For the version positioning in the nav, this can be switched from right to left within the `docusaurus.config.js` file on line 251. At the moment it's set to right, but can be moved to the left by updating the setting there

**Comment by jtcohen6** [comment]
> In order to share a version-specific url, this can be done by adding a version param to the url, for example: https://deploy-preview-1052--docs-getdbt-com.netlify.app/?version=0.21

Makes sense! Would it be possible to update/store the value in the URL every time someone selects a version option? (Joel - really good catch - this is so so so important for troubleshooting/support)

**Comment by joellabes** [comment]
Should the picker show when on the Dev blog? feels a bit strange given that changing it won't do anything: https://deploy-preview-1052--docs-getdbt-com.netlify.app/blog/customer-360-view-identity-resolution 

Also, the version takes a while to populate in the header when it's accessed by URL: https://www.loom.com/share/c411cab8d63b41c3a9e4e048f0f5c646

File: website/dbt-global-variables.js
Comment (runleonarun) [reply website/dbt-global-variables.js]: @joellabes It is a good way to easily rename a feature without the hassle of missing many instances like we would with search & replace. But more importantly, it's a way to standardize on spelling and capitalization of a feature or any name in dbt. A great example is `dbt Labs` If we use a product variable for when we use the company name, then it will decrease the chance there will be typos (dBt labs`</gasp>`). 

Product variables will be really important as we scale and more folks are writing docs. For example, do we say dbt Core v1 or dbt Core v1.0 or dbt Core Version 1.0 ?!? Product variables will also be useful when we're ready to translate. The translators can ignore everything in a variable OR translate in one place, consistently.

File: website/src/theme/DocPage/styles.module.css
Comment (runleonarun) [reply website/src/theme/DocPage/styles.module.css]: We did some more color spelunking today and Jason is going to make some changes to improve the notes and banners.

File: website/dbt-versions.js
Comment (JKarlavige) [reply website/dbt-versions.js]: The latest two commits add this feature, with the EOD date being set in the `dbt-versions` file.

**Comment by JKarlavige** [comment]
> Should the picker show when on the Dev blog? feels a bit strange given that changing it won't do anything: https://deploy-preview-1052--docs-getdbt-com.netlify.app/blog/customer-360-view-identity-resolution
> 
> Also, the version takes a while to populate in the header when it's accessed by URL: https://www.loom.com/share/c411cab8d63b41c3a9e4e048f0f5c646

Thanks for sharing this @joellabes, I don't think we need to show the version nav item if in the blog directory. @runleonarun what are your thoughts with this?

For the version nav item, I will hide the version nav item entirely until the data has loaded to clean this up a bit.

**Comment by JKarlavige** [comment]
> > In order to share a version-specific url, this can be done by adding a version param to the url, for example: https://deploy-preview-1052--docs-getdbt-com.netlify.app/?version=0.21
> 
> Makes sense! Would it be possible to update/store the value in the URL every time someone selects a version option? (Joel - really good catch - this is so so so important for troubleshooting/support)

This is a great point. At the moment it is not setup to add the version to the URL, however I feel that this _should_ be possible. Will add it to the asana task to look into.

**Comment by JKarlavige** [comment]
> > > In order to share a version-specific url, this can be done by adding a version param to the url, for example: https://deploy-preview-1052--docs-getdbt-com.netlify.app/?version=0.21
> > 
> > 
> > Makes sense! Would it be possible to update/store the value in the URL every time someone selects a version option? (Joel - really good catch - this is so so so important for troubleshooting/support)
> 
> This is a great point. At the moment it is not setup to add the version to the URL, however I feel that this _should_ be possible. Will add it to the asana task to look into.

Hey @jtcohen6, from looking into this further I don't believe it's something that will fall within the realm of the initial scope of this task. I wouldn't rule it out entirely, but it may be better fit for a "v2" round of updates for the versioning feature. 

I'm hesitant to attempt to fit this into the initial release, as it would be a global change affecting every URL. This is something I'd want to separate out into it's own project to see what options there would be and spend a good amount of time working on and testing. It may involve tapping into the Docusaurus routing to append the version onto every URL.

I believe @runleonarun has a list of potential features to add into this project. We should definitely note this as a higher-priority item to circle back to after we go live.

**Comment by runleonarun** [comment]
> I believe @runleonarun has a list of potential features to add into this project. We should definitely note this as a higher-priority item to circle back to after we go live.

@JKarlavige Yep! I just [added it here](https://www.notion.so/dbtlabs/Docs-versioning-project-3f670deb35fe4df7bf8e184363b2dfb1#ab5b9d0f837f4b5291918b0c4e30a002). and I will create a GitHub issue to track Versioning V2.

File: website/src/theme/DocPage/index.js
Comment (runleonarun) [inline website/src/theme/DocPage/index.js]: @runleonarun revise this banner text!

File: website/src/theme/DocPage/index.js
Comment (runleonarun) [inline website/src/theme/DocPage/index.js]: @runleonarun review this text!

File: website/src/theme/DocPage/index.js
Comment (runleonarun) [inline website/src/theme/DocPage/index.js]: @runleonarun review this text

File: website/snippets/test-snippet.md
Comment (runleonarun) [inline website/snippets/test-snippet.md:5]: @runleonarun Use real life snippet!

File: website/dbt-versions.js
Comment (runleonarun) [inline website/dbt-versions.js:11]: @runleonarun put real life EOL scenarios in here. 

cc @jtcohen6 for awareness

File: website/src/theme/DocPage/styles.module.css
Comment (runleonarun) [reply website/src/theme/DocPage/styles.module.css]: We settled on brand colors and got approval form Design!

File: website/dbt-versions.js
Comment (runleonarun) [inline website/dbt-versions.js]:
```suggestion
    version: "1.1",
```

**Comment by JKarlavige** [comment]
Sharing here for visibility. I have the first rough draft for documentation on this feature together here: https://www.notion.so/dbtlabs/Versioning-on-Docusaurus-c6a4a41a66cd4ea2970854cc42cb5b70

File: website/docs/reference/node-selection/defer.md
Comment (runleonarun) [reply website/docs/reference/node-selection/defer.md]: @JKarlavige Here are some versioning clarifications that I discussed with Jeremy this week. I also captured them in our [notes for the week](https://www.notion.so/dbtlabs/March-8th-2022-26f5c53fe8864843974e3767a4bac31b).
    * We will not be surfacing patch version number or x. So, use 1.1 and not 1.1.x. We will message this in the Core versioning policy, which I captured in a follow-up task https://github.com/dbt-labs/docs.getdbt.com/issues/1217
    * Versioning in the picker will include `.20`,`.21`, `1.0` and higher. BUT we will remove `.20` and `.21` on June 30th, and communicate this in a banner that says they will no longer be supported after 6/30 and in the core versioning policy page. Also captured in a follow-up task https://github.com/dbt-labs/docs.getdbt.com/issues/1217
   
Let me know if there's anything else we can clarify!

File: website/dbt-versions.js
Comment (runleonarun) [reply website/dbt-versions.js]: @JKarlavige Should I comment this out until the 1.1 release?

Comment (JKarlavige) [reply website/dbt-versions.js]: @runleonarun committed your suggestion and commented out 1.1!

**Comment by JKarlavige** [comment]
Hey @jtcohen6 and @joellabes, thank you for your feedback and input so far on this project. We are aiming to take this live a week from now on March 17th. There is still a few places with placeholder data we need to swap out, but a majority of the functionality is ready to go. I opened this PR for a final run-through and review, and any help with the review would be greatly appreciated!

@lfracalossiFA adding you for review as well

**Comment by runleonarun** [comment]
@amychen1776 I wonder if you'd also want to take a look at these upcoming changes to let us know if you notice anything. Jason linked this above, but it's worth relinking the [documentation he wrote up](https://www.notion.so/dbtlabs/Versioning-on-Docusaurus-c6a4a41a66cd4ea2970854cc42cb5b70) for us to use the new features.

File: website/dbt-versions.js
Comment (runleonarun) [inline website/dbt-versions.js:12]:
```suggestion
    EOLDate: "2022-06-30"
```

File: website/dbt-versions.js
Comment (runleonarun) [inline website/dbt-versions.js:16]:
```suggestion
    EOLDate: "2022-06-30"
```

Comment (runleonarun) [inline website/dbt-versions.js]:
```suggestion
    EOLDate: "2022-06-30"
```

File: website/dbt-versions.js
Comment (runleonarun) [inline website/dbt-versions.js]:
```suggestion
    EOLDate: "2022-06-30"
```

File: website/dbt-versions.js
Comment (runleonarun) [inline website/dbt-versions.js]:
```suggestion
    EOLDate: "2022-12-03"
```

File: website/dbt-versions.js
Comment (runleonarun) [inline website/dbt-versions.js]: @jtcohen6 I think we need this to be available for the beta, right?
```suggestion
  //   EOLDate: "2023-03-18"
```

**Comment by lfracalossiFA** [comment]
I have a UX question, really for the whole team. I went to the [URL Jason put in the PR description](https://deploy-preview-1052--docs-getdbt-com.netlify.app/reference/node-selection/defer), and it's unclear to me what version I'm actually looking at here. It says "v1.0" in the dropdown, but it also says "This version is nearing the end of support. The latest stable version is 1.0." So am I not in fact viewing 1.0 docs when I look at that page?

Could this be confusing to users? Maybe I'm missing something that would be obvious to dbt power users, though.

**Comment by JKarlavige** [comment]
> 

This is due to us still having placeholder dates for each version. @runleonarun has a note to update those to the correct dates. In the meantime, I will adjust that date forward a year to reflect more how it would actually look when live!

File: website/dbt-versions.js
Comment (runleonarun) [inline website/dbt-versions.js]: @jtcohen6 how long do you want to have 1.0 show up in the docs?
```suggestion
    EOLDate: "2023-12-03"
```

File: website/src/theme/DocPage/index.js
Comment (runleonarun) [inline website/src/theme/DocPage/index.js]:
```suggestion
        EOLBannerText: `This version of dbt Core is nearing the end of its [critical support period](/docs/core-versions). For better performance, improved security, and new features, you should upgrade to ${latestStableRelease}, the latest stable version.`
```

File: website/src/theme/DocPage/index.js
Comment (runleonarun) [inline website/src/theme/DocPage/index.js]:
```suggestion
                  <p> You should upgrade to {firstAvailableVersion} or later if you want to use this feature.</p>
```

File: website/src/theme/DocPage/index.js
Comment (runleonarun) [inline website/src/theme/DocPage/index.js]:
```suggestion
                  <p style={{'marginTop': '5px', 'marginBottom': '0'}}>Unfortunately, this feature is not available in dbt Core version {dbtVersion}</p>
```

File: website/src/theme/DocPage/index.js
Comment (runleonarun) [inline website/src/theme/DocPage/index.js]:
```suggestion
        EOLBannerText: `This version of dbt Core is [no longer supported](/docs/core-versions). No patch releases will be made, even for critical security issues. For better performance, improved security, and new features, you should upgrade to ${latestStableRelease}, the latest stable version.`
```

File: website/src/theme/DocPage/index.js
Comment (runleonarun) [reply website/src/theme/DocPage/index.js]: @JKarlavige Can we add the the EOL date do we already have that as a variable?

**Review by runleonarun — APPROVED** [review]
@JKarlavige and @jtcohen6 I revised the support notes and added real life version scenarios.  Let me know if you feel like any of it should change!

Jason, I was wondering if I could add the EOL date in the note too.

File: website/src/theme/DocPage/index.js
Comment (JKarlavige) [reply website/src/theme/DocPage/index.js]: @runleonarun Yes, the EOL date is already available using the `EOLDate` variable, in the same way we add the `latestStableRelease` variable to the banner text. The EOL date format is currently how we set it in the dbt-versions file: `2023-05-20`. Let me know if you prefer to display the date differently and I will push an update.

**Comment by JKarlavige** [comment]
@runleonarun Committed the suggestions made for the content in the banners. Holding off on committing the EOLDate updates. Once you and @jtcohen6 have those sorted out, feel free to commit them!

**Review by lfracalossiFA — APPROVED** [review]
Looks good to me! There is still a test snippet that it looks like Leona is working on still, but I assume you are already aware of it.

**Comment by JKarlavige** [comment]
Put together a [Loom video](https://www.loom.com/share/514bcd6f62b4483e88a662003d555e3f) to walk through how the end-of-life dates, and versioned pages work. These won't be visible in the preview, as our current dates set won't trigger these banners.

**Comment by JKarlavige** [comment]
> 

Awesome! And yep Leona is leaving that in place, and will being adding actual snippets down the road. It's okay to have that demo snippet, as long as we're not actually using that demo snippet within a `<Snippet>` component anywhere on a page.

**Review by breezyfasano — APPROVED** [review]
lgtm!!! great work.

File: website/dbt-versions.js
Comment (jtcohen6) [reply website/dbt-versions.js]: ✅ correct!

File: website/dbt-versions.js
Comment (jtcohen6) [inline website/dbt-versions.js:16]: This feels like a fine middle ground for now. @runleonarun and I discussed this at length in our last 1:1.

My main concerns are:
- These versions have already reached EOL from a Core maintenance perspectively, but they'll continue to be supported in Cloud until June 30.
- The selection of v0.20 or v0.21 from the version dropdown implies that the user will see docs that are fully relevant for v0.20 or v0.21, which isn't quite true. We'll add some language to the `Core Versions` page to reflect this: [EOLDate](https://github.com/dbt-labs/docs.getdbt.com/issues/1217)

Ideally, we'd have a banner that we can show users for the next 3 months, while v0.20 + v0.21 are special exceptions. I don't think that should block shipping all the good work that's gone into this feature, and which we badly need to unblock other initiatives.

**Review by jtcohen6 — APPROVED** [review]
Let's ship it! Great work on this @JKarlavige @runleonarun

File: website/dbt-versions.js
Comment (JKarlavige) [reply website/dbt-versions.js:16]: For sure! This initial release will be minimum as far as actual versioned content goes. The only versioned content this PR currently includes come from the Node Selector > Commands pages, converting the tabs to the versionBlock component. 

I can imagine the more versioned content we add, the more we will want to adjust. This is a great item to add to our list for continuous improvements on this project.

**Pull request merged**
